### PR TITLE
cmake: Add support for variable offset and slot size

### PR DESCRIPTION
**Pull request by sigvartmh**
To support the use case of multi-image MCUBoot the slot size and offset
of the signing proceedure needs to be configurable. This commit will add
these as variables to be put into the signing command.

It also shares the variable `CONFIG_UPDATEABLE_IMAGE_NUMBER` from
mcuboot to other images so that it can be used for setting different
configurations and sizes.

This commit also adds the generation of a dfu zip file with manifest for
network and application core when building for nRF53.

Signed-off-by: Sigvart Hovland <sigvart.hovland@nordicsemi.no>

### CODE REVIEW

**Comment by sigvartmh** [comment]
Please tell me if you feel like I should split this up into multiple commits.

File: cmake/partition_manager.cmake [hunk 1]
```diff
@@ -185,6 +185,19 @@ if (CONFIG_PM_EXTERNAL_FLASH)
     )
 endif()
 
+if (CONFIG_NRF53_UPGRADE_NETWORK_CORE
+  AND CONFIG_HCI_RPMSG_BUILD_STRATEGY_FROM_SOURCE
+  AND (mcuboot_UPDATEABLE_IMAGE_NUMBER GREATER 1))
+  add_region(
+    NAME ram_flash
+    SIZE 0x40000
```
Comment (sigvartmh) [inline cmake/partition_manager.cmake:193]: These values should ideally come from the DT node of flash_ctrl if possible.

File: cmake/partition_manager.cmake [hunk 2]
```diff
@@ -185,6 +185,19 @@ if (CONFIG_PM_EXTERNAL_FLASH)
     )
 endif()
 
+if (CONFIG_NRF53_UPGRADE_NETWORK_CORE
+  AND CONFIG_HCI_RPMSG_BUILD_STRATEGY_FROM_SOURCE
+  AND (mcuboot_UPDATEABLE_IMAGE_NUMBER GREATER 1))
+  add_region(
+    NAME ram_flash
```
Comment (sigvartmh) [inline cmake/partition_manager.cmake:192]: This is adding an emulated flash region which is also defined in DTS

File: cmake/multi_image.cmake [hunk 3]
```diff
@@ -29,6 +29,9 @@ if(IMAGE_NAME)
   share("list(APPEND ${IMAGE_NAME}BUILD_BYPRODUCTS ${PROJECT_BINARY_DIR}/${KERNEL_ELF_NAME})")
   # Share the signing key file so that the parent image can use it to
   # generate signed update candidates.
+  if(CONFIG_UPDATEABLE_IMAGE_NUMBER)
+    share("set(${IMAGE_NAME}UPDATEABLE_IMAGE_NUMBER ${CONFIG_UPDATEABLE_IMAGE_NUMBER})")
```
Comment (sigvartmh) [inline cmake/multi_image.cmake:33]: @tejlmand  since https://github.com/zephyrproject-rtos/zephyr/blob/7c2dfc1558e598e07c015738e984f5c50c3e6c82/subsys/dfu/boot/mcuboot.c#L19 has started to use the bootutil_public API we need a way of sharing configuration between MCUBoot and Application since changes in MCUBoot config are not propagated to the application sourcing of `mcuboot_config.h` here https://github.com/mcu-tools/mcuboot/blob/78e4441bd373b2ea8a7c36787ee657953c65481f/boot/bootutil/include/bootutil/bootutil_public.h#L44

As you can see in 
https://github.com/mcu-tools/mcuboot/blob/main/boot/zephyr/include/mcuboot_config/mcuboot_config.h
there area lot of configurations sourced from the Kconfig which are not reflected in the application sourcing of `mcuboot_config.h` since they will not be defined as they do not exists.

I don't know if you have any good idea how to do this or if we should just try to keep adding the configurations we need to be set. 

For multi-image applications that would making sure `CONFIG_UPDATEABLE_IMAGE_NUMBER > 1` but also we need the coming  hooks API(https://github.com/mcu-tools/mcuboot/pull/1058). Where we need to make sure `MCUBOOT_IMAGE_ACCESS_HOOKS` is set by propagating the configuration.

1. Solution is to mirror the Kconfigs and set them also in application image.

File: cmake/partition_manager.cmake (hunk 1 above)
Comment (hakonfam) [reply cmake/partition_manager.cmake:193]: You already have a kconfig symbol for the FLASH size that you can use `CONFIG_FLASH_SIZE` from `zephyr/arch/kconfig`

File: cmake/partition_manager.cmake [hunk 4]
```diff
@@ -185,6 +185,19 @@ if (CONFIG_PM_EXTERNAL_FLASH)
...
+    SIZE 0x40000
+    BASE 0x00000000
```
Comment (hakonfam) [inline cmake/partition_manager.cmake:194]: The base address also has a kconfig symbol. Note that both of these are derived from DT

File: scripts/bootloader/generate_zip.py [hunk 5]
```diff
@@ -63,8 +63,8 @@ def get_name(text_wrapper):
         special_info[n]['file'] = n
         special_info[n]['modtime'] = int(path.getmtime(p))
         merged = dict()
-        merged.update(special_info[n])
         merged.update(shared_info)
+        merged.update(special_info[n])
```
Comment (hakonfam) [inline scripts/bootloader/generate_zip.py:67]: Why is this moved?

File: samples/CMakeLists.txt [hunk 6]
```diff
@@ -82,6 +82,8 @@ endif()
 
 if (CONFIG_SECURE_BOOT)
   if (CONFIG_SOC_NRF5340_CPUNET)
+    share("set(${IMAGE_NAME}soc ${CONFIG_SOC})")
```
Comment (hakonfam) [inline samples/CMakeLists.txt:85]: Why do we need to share this?

Comment (hakonfam) [reply samples/CMakeLists.txt:85]: Please add a comment explaining why (creating signed zips)

File: cmake/partition_manager.cmake [hunk 7]
```diff
@@ -437,8 +450,66 @@ else()
...
           # There is no padding in front of the network core application.
-          math(EXPR net_app_TO_SECONDARY
-            "${PM_MCUBOOT_SECONDARY_ADDRESS} - ${net_app_addr} + ${PM_MCUBOOT_PAD_SIZE}")
+	  if(CONFIG_PM_EXTERNAL_FLASH)
+	    if(CONFIG_SOC_NRF5340_CPUAPP)
+	      set(EXTERNAL_FLASH_START_ADDRESS 0x10000000)
+            else()
+	      set(EXTERNAL_FLASH_START_ADDRESS 0x12000000)
+            endif()
+            math(EXPR net_app_TO_SECONDARY "${EXTERNAL_FLASH_START_ADDRESS} + ${CONFIG_PM_EXTERNAL_FLASH_BASE}
```
Comment (hakonfam) [inline cmake/partition_manager.cmake:459]: All of these calculations looks complicated.

I would suggest extracting this to a separate file and adding comments to all the calculations explaining why they are needed.

File: cmake/partition_manager.cmake [hunk 8]
```diff
@@ -437,8 +450,66 @@ else()
             )
 
           # There is no padding in front of the network core application.
-          math(EXPR net_app_TO_SECONDARY
-            "${PM_MCUBOOT_SECONDARY_ADDRESS} - ${net_app_addr} + ${PM_MCUBOOT_PAD_SIZE}")
+	  if(CONFIG_PM_EXTERNAL_FLASH)
+	    if(CONFIG_SOC_NRF5340_CPUAPP)
+	      set(EXTERNAL_FLASH_START_ADDRESS 0x10000000)
```
Comment (hakonfam) [inline cmake/partition_manager.cmake:455]: These hard coded symbols should be prompt-less kconfig-options IMO

**Comment by sigvartmh** [comment]
I'll change to using the - set_shared(IMAGE <name> PROPERTY <prop> <value>)
- get_shared(<var> IMAGE <name> PROPERTY <prop>) and rebase it
-

File: cmake/partition_manager.cmake (hunk 7 above)
Comment (tejlmand) [reply cmake/partition_manager.cmake:459]: All those calculations looks similar, is it not possible to just make a single function or macro with inputs, something like:
```
# secondary_slot([EXT_START <address>] [EXT_BASE <address>])
function(secondary_slot [EXT_START <address>] [EXT_BASE <address>] )
  cmake_parse_arguments(SEC_SLOT "" "EXT_START;EXT_BASE" "" ${ARGN}
  get_shared(no IMAGE mcuboot PROPERTY UPDATEABLE_IMAGE_NUMBER)

  if(${no} GREATER 1)
    set(sec_slot_idx "_1")
  endif()

  get_target_property(
    net_app_addr
    partition_manager
    CPUNET_PM_APP_ADDRESS
  )
  
  math(EXPR net_app_to_secondary
    "${SEC_SLOT_EXT_START} + ${SEC_SLOT_EXT_BASE} + ${PM_MCUBOOT_SECONDARY${sec_slot_idx}_ADDRESS} - ${net_app_addr} + ${PM_MCUBOOT${sec_slot_idx}_PAD_SIZE}")
  
  set_property(
    TARGET partition_manager
    PROPERTY net_app_TO_SECONDARY
    ${net_app_TO_SECONDARY}
  )  
endfunction()
``` 

and then change all the if'es to:
```
if(CONFIG_PM_EXTERNAL_FLASH)
  secondary_slot(EXT_START ${EXTERNAL_FLASH_START_ADDRESS} EXT_BASE ${CONFIG_PM_EXTERNAL_FLASH_BASE})
else()
  secondary_slot()
endif()
```
as `UPDATEABLE_IMAGE_NUMBER` is fetched inside the function.
Might even be improved further, and obtain better naming.

File: modules/mcuboot/CMakeLists.txt [hunk 9]
```diff
@@ -16,15 +16,16 @@ if(CONFIG_BOOTLOADER_MCUBOOT)
     # SIGNED_HEX_FILE_OUT - (required) Signed hex output image
     # DEPENDS - (optional) One or more dependencies for signing the hex image
 
-    set(oneValueArgs SIGNED_BIN_FILE_IN SIGNED_HEX_FILE_NAME_PREFIX START_ADDRESS_OFFSET SIGNED_HEX_FILE_OUT)
+    set(oneValueArgs SIGNED_BIN_FILE_IN SIGNED_HEX_FILE_NAME_PREFIX SLOT_SIZE START_ADDRESS_OFFSET SIGNED_HEX_FILE_OUT)
```
Comment (tejlmand) [inline modules/mcuboot/CMakeLists.txt:19]: SLOT_SIZE not documented.

And why not just size, is it important for the sign command to know that it's a slot size ?
I would assume it just signs according to the size given.

File: modules/mcuboot/CMakeLists.txt [hunk 10]
```diff
@@ -16,15 +16,16 @@ if(CONFIG_BOOTLOADER_MCUBOOT)
...
 
     if (NOT DEFINED SIGN_ARG_SIGNED_BIN_FILE_IN
         OR NOT DEFINED SIGN_ARG_SIGNED_HEX_FILE_OUT
         OR NOT DEFINED SIGN_ARG_SIGNED_HEX_FILE_NAME_PREFIX
-        OR NOT DEFINED SIGN_ARG_START_ADDRESS_OFFSET)
-      message(FATAL_ERROR "Missing parameter, required: SIGNED_BIN_FILE_IN SIGNED_HEX_FILE_NAME_PREFIX START_ADDRESS_OFFSET SIGNED_HEX_FILE_OUT")
+        OR NOT DEFINED SIGN_ARG_START_ADDRESS_OFFSET
+        OR NOT DEFINED SIGN_ARG_SLOT_SIZE)
+      message(FATAL_ERROR "Missing parameter, required: SIGNED_BIN_FILE_IN SIGNED_HEX_FILE_NAME_PREFIX START_ADDRESS_OFFSET SIGNED_HEX_FILE_OUT SLOT_SIZE")
     endif()
```
Comment (tejlmand) [inline modules/mcuboot/CMakeLists.txt:29]: Consider changing this code block to:
https://github.com/nrfconnect/sdk-nrf/blob/d34214d69056671a3d9184332c139f98edbebff0/cmake/extensions.cmake#L39
```suggestion
  check_arguments_required_all(sign SIGN_ARG ${oneValueArgs})
```

File: modules/mcuboot/CMakeLists.txt [hunk 11]
```diff
@@ -159,6 +164,8 @@ if(CONFIG_BOOTLOADER_MCUBOOT)
   endif()
 
   if(CONFIG_SIGN_IMAGES)
+    set(app_core_binary_name app_update.bin)
```
Comment (tejlmand) [inline modules/mcuboot/CMakeLists.txt:167]: Should `app_update.bin` be configurable or maybe `${CONFIG_KERNEL_BIN_NAME}_update.bin` ?
So that it follows the app name, if user changes from `zephyr.elf` to `awesome_app.elf`, then the update image follow suit and becomes `awesome_app_update.bin`.

File: modules/mcuboot/CMakeLists.txt [hunk 12]
```diff
@@ -264,6 +264,32 @@ if(CONFIG_BOOTLOADER_MCUBOOT)
...
+    else()
+      generate_dfu_zip(
+        TARGET mcuboot_sign_target
+        OUTPUT ${PROJECT_BINARY_DIR}/dfu_application.zip
+        BIN_FILES ${PROJECT_BINARY_DIR}/${app_core_binary_name}
+        TYPE application
+        SCRIPT_PARAMS
+        "load_address=$<TARGET_PROPERTY:partition_manager,PM_APP_ADDRESS>"
+        "version_MCUBOOT=${CONFIG_MCUBOOT_IMAGE_VERSION}"
+        )
```
Comment (tejlmand) [inline modules/mcuboot/CMakeLists.txt:292]: could you try to combine to a single `generate_dfu_zip()` call ?

**Review by tejlmand — CHANGES_REQUESTED** [review]
CMake code needs cleanup, see individual comments.

File: cmake/partition_manager.cmake (hunk 8 above)
Comment (sigvartmh) [reply cmake/partition_manager.cmake:455]: Unfortunately, these values are not found in the device tree. They are currently fetched from https://infocenter.nordicsemi.com/topic/ps_nrf5340/chapters/memory/appmem.html?cp=3_0_0_4_2 and 
https://infocenter.nordicsemi.com/topic/ps_nrf52840/memory.html?cp=4_0_0_3_1_2#memorymap

IMO they should have been encoded in the DTS of the SoC but I couldn't find a suitable node.

File: cmake/partition_manager.cmake (hunk 1 above)
Comment (sigvartmh) [reply cmake/partition_manager.cmake:193]: CONFIG_FLASH_SIZE will be wrong it needs to be CPUNET_FLASH_SIZE

File: cmake/partition_manager.cmake (hunk 2 above)
Comment (sigvartmh) [reply cmake/partition_manager.cmake:192]: Will do

File: cmake/partition_manager.cmake (hunk 4 above)
Comment (sigvartmh) [reply cmake/partition_manager.cmake:194]: The base address used here is fake since `flash_simulator` does not use a real base address since it stores things in a global array called `mock_flash`. I'll add Kconfig with hooks into the `flash_ctrl` device tree node.

File: scripts/bootloader/generate_zip.py (hunk 5 above)
Comment (sigvartmh) [reply scripts/bootloader/generate_zip.py:67]: This change is propagated from https://github.com/nrfconnect/sdk-nrf/pull/5206/commits/9afe7a5f0133da29e2b2e02d9f48ff38c627be35 I don't know.

File: modules/mcuboot/CMakeLists.txt (hunk 9 above)
Comment (sigvartmh) [reply modules/mcuboot/CMakeLists.txt:19]: The sign command needs to know the slot size unfortunately as it puts some data at the end of the slot.
So the size of the image can be smaller than the slot size, but the sign command needs to know where it should offset the TLV data at the end of the image.

File: modules/mcuboot/CMakeLists.txt (hunk 11 above)
Comment (hakonfam) [reply modules/mcuboot/CMakeLists.txt:167]: I agree that this is better, but I'm afraid of breaking compatibility with tools that depend on the name of the generated update bin

File: modules/mcuboot/CMakeLists.txt (hunk 9 above)
Comment (tejlmand) [reply modules/mcuboot/CMakeLists.txt:19]: Then `SLOT_SIZE` does sound like the best name.
But it is still not documented, please add description of SLOT_SIZE here:
https://github.com/nrfconnect/sdk-nrf/blob/4af2c283eea8e2a1d5f7663a0ebe659e2cbd7d82/modules/mcuboot/CMakeLists.txt#L12-L17

Please provide the same details as in your last comment, so readers understand the difference between image size and slot size.

File: modules/mcuboot/CMakeLists.txt (hunk 11 above)
Comment (tejlmand) [reply modules/mcuboot/CMakeLists.txt:167]: could it be investigated whether this is the case.
Doesn't the zip contain a meta file that describes exactly which file in the zip that goes to which core, so the tools are using that ?

Comment (tejlmand) [reply modules/mcuboot/CMakeLists.txt:167]: To me it looks as if the file name is being passed on from CMake and added to a manifest here:
https://github.com/nrfconnect/sdk-nrf/blob/4af2c283eea8e2a1d5f7663a0ebe659e2cbd7d82/cmake/fw_zip.cmake#L27
https://github.com/nrfconnect/sdk-nrf/blob/4af2c283eea8e2a1d5f7663a0ebe659e2cbd7d82/scripts/bootloader/generate_zip.py#L63

Disclaimer, there could be details in this i'm not familiar with.

Comment (hakonfam) [reply modules/mcuboot/CMakeLists.txt:167]: I will double check, if its the case then I will change it.
